src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala

```diff
@@ -771,63 +774,73 @@ class ILoop(config: ShellConfig, inOverride: BufferedReader = null,
-        (raw0, Option(file0), Option(margin0))
+    val (flags, args) = tokenize(arg).span(_.startsWith("-"))
+    def raw  = flags.contains("-raw")
+    def java = flags.contains("-java")
+    def usage() = echo("usage: :paste [-raw | -java] file | < EOF")
+    def pasteFile(name: String): String = {
+      label = name
+      withFile(name) { f =>
+        shouldReplay = Some(s":paste $arg")
+        f.slurp().trim().tap(s => echo(if (s.isEmpty) s"File contains no code: $f" else s"Pasting file $f..."))
+      }.getOrElse("")
+    }
+    def pasteWith(margin: String, eof: Option[String]): String = {
+      echo(s"// Entering paste mode (${ eof getOrElse "ctrl-D" } to finish)\n")
+      in.withSecondaryPrompt("") {
+        val delimiter = eof.orElse(config.pasteDelimiter.option)
+        def atEOF(s: String) = delimiter.map(_ == s).getOrElse(false)
+        val input = readWhile(s => !atEOF(s)).mkString("\n")
+        val text =
+          margin match {
+            case ""  => input.trim
+            case "-" => input.linesIterator.map(_.trim).mkString("\n")
+            case _   => input.stripMargin(margin.head).trim
+          }
+        echo(if (text.isEmpty) "\n// Nothing pasted, nothing gained.\n" else "\n// Exiting paste mode, now interpreting.\n")
+        text
       }
-    val code = (file, margin) match {
-      case (Some(name), None) =>
-        label = name
-        withFile(name) { f =>
-          shouldReplay = Some(s":paste $arg")
-          val s = f.slurp().trim()
-          if (s.isEmpty) echo(s"File contains no code: $f")
-          else echo(s"Pasting file $f...")
-          s
-        } getOrElse ""
-      case (eof, _) =>
-        echo(s"// Entering paste mode (${ eof getOrElse "ctrl-D" } to finish)\n")
-        in.withSecondaryPrompt("") {
-          val delimiter = eof orElse config.pasteDelimiter.option
-          val input = readWhile(s => delimiter.isEmpty || delimiter.get != s) mkString "\n"
-          val text = (
-            margin filter (_.nonEmpty) map {
-              case "-" => input.linesIterator map (_.trim) mkString "\n"
-              case m => input stripMargin m.head // ignore excess chars in "<<||"
-            } getOrElse input
-            ).trim
-          if (text.isEmpty) echo("\n// Nothing pasted, nothing gained.\n")
-          else echo("\n// Exiting paste mode, now interpreting.\n")
-          text
-        }
     }
-    def interpretCode() = {
-      if (intp.withLabel(label)(intp interpret code) == Incomplete)
-        paste.incomplete("The pasted code is incomplete!\n", label, code)
+    val code: String = args match {
+      case name :: Nil if !name.startsWith("<")       => pasteFile(name)
+      case Nil                                        => pasteWith("", None)
+      case here :: Nil                                => pasteWith(here.slice(1, 2), None)
+      case here :: eof :: Nil if here.startsWith("<") => pasteWith(here.slice(1, 2), Some(eof))
+      case _ => usage() ; ???
```

Comment:
looks like a forgotten `???`?  It used to early return before.

Reply:
`s/forgotten/suppressed` I assume at least one follow-up to make it useful.

Reply:
Only for experimental things.  But I wouldn't want to accidentally ship `:paste` broken by merging this as such.

Reply:
But who isn't broken? Who isn't accidentally shipped? What is birth but an accidental shipment?

I will WIPpify this PR.

Reply:
Ah, I misread your "follow-up" as in a follow-up PR from merging this one.